src/smt.rs

```diff
@@ -473,6 +508,246 @@ impl BaseSort for SMTSolved {
+                .clone(),
+        });
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum SMTUFIntValue {
+    Declaration(String, Vec<String>),
+}
+
+impl SMTUFIntValue {
+    pub fn to_uf<'s>(&self, st: &'s Storage) -> Fun<'s> {
+        match self {
+            SMTUFIntValue::Declaration(name, types) => {
+                let sorts: Result<Vec<Sort>, String> = types
+                    .iter()
+                    .map(|type_name| match type_name.as_str() {
+                        "Int" => Ok(Int::sort()),
+                        "Bool" => Ok(Bool::sort()),
+                        other => Err(format!("unknown type {other}")),
+                    })
+                    .collect();
+                Fun::new(st, name.to_string(), sorts.unwrap(), Int::sort())
+            }
+        }
+    }
+
+    pub fn to_term(&self, termdag: &mut TermDag) -> Term {
+        match self {
+            SMTUFIntValue::Declaration(name, types) => {
+                let name = termdag.lit(Literal::String(name.clone()));
+                let mut children: Vec<Term> = types
+                    .iter()
+                    .map(|type_name| termdag.lit(Literal::String(type_name.clone())))
+                    .collect();
+                children.insert(0, name);
+                termdag.app("smt-fn-int".into(), children)
+            }
+        }
+    }
+}
+
+impl BaseValue for SMTUFIntValue {}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum SMTBaseValue {
+    BoolValue(SMTBoolValue),
+    IntValue(SMTIntValue),
+    RealValue(SMTRealValue),
+}
+
+impl BaseValue for SMTBaseValue {}
+
+impl SMTBaseValue {
+    pub fn to_term(&self, termdag: &mut TermDag) -> Term {
+        match self {
+            SMTBaseValue::BoolValue(arg) => arg.to_term(termdag),
+            SMTBaseValue::IntValue(arg) => arg.to_term(termdag),
+            SMTBaseValue::RealValue(arg) => arg.to_term(termdag),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct SMTUFInt;
+impl BaseSort for SMTUFInt {
+    type Base = SMTUFIntValue;
+
+    fn name(&self) -> &str {
+        "SMTUFInt"
+    }
+
+    fn reconstruct_termdag(
+        &self,
+        base_values: &BaseValues,
+        value: Value,
+        termdag: &mut TermDag,
+    ) -> Term {
+        let uf = base_values.unwrap::<SMTUFIntValue>(value);
+        uf.to_term(termdag)
+    }
+
+    fn register_primitives(&self, eg: &mut EGraph) {
+        // (smt-uf-int "f" "Int" "Int")
+        add_primitive!(
+            eg,
+            "smt-fn-int" = [args: S] -> SMTUFIntValue { {
+                let mut args_iter = args.into_iter();
+                let name = args_iter.next().unwrap();
+                let types: Vec<String> = args_iter.map(Boxed::into_inner).collect();
+                SMTUFIntValue::Declaration(name.0, types)
+            } }
+        );
+    }
+}
+
+#[derive(Clone)]
+struct UFApply {
+    args: Vec<ArcSort>,
+    out_sort: ArcSort,
+    fun_sort: ArcSort,
```

Comment:
We just chatted about this, for other types likes reals and bools we can make a new instance of this struct to make return type specific versions of each and register them as new primitives, but all with the same name (which should be fine, since type checking will know)